File: src/cn/vove7/plugin/rest/filetype/RequestParser.java
package cn.vove7.plugin.rest.filetype;


import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;

import cn.vove7.plugin.rest.model.RequestModel;
import cn.vove7.plugin.rest.psi.RestEHeader;
import cn.vove7.plugin.rest.psi.RestEMethod;
import cn.vove7.plugin.rest.psi.RestEParam;
import cn.vove7.plugin.rest.psi.RestEUrl;
import cn.vove7.plugin.rest.psi.RestHeaders;
import cn.vove7.plugin.rest.psi.RestParams;
import cn.vove7.plugin.rest.psi.RestRequestBody;

/**
 * @author danblack
 */
public class RequestParser {

    private static RequestModel.Method getMethod(cn.vove7.plugin.rest.psi.RestRequest request) {
        RestEMethod method = request.getEMethod();
        if (method != null && method.isValid()) {
            switch (method.getText()) {
                case "GET":
                    return RequestModel.Method.GET;
                case "PUT":
                    return RequestModel.Method.PUT;
                case "POST":
                    return RequestModel.Method.POST;
                case "DELETE":
                    return RequestModel.Method.DELETE;
                case "PATCH":
                    return RequestModel.Method.PATCH;
            }
            throw new IllegalStateException();
        }
        return RequestModel.Method.GET;
    }

    public static RequestModel parse(cn.vove7.plugin.rest.psi.RestRequest request) {
        RequestModel.Method method = getMethod(request);
        String url = getUrl(request);
        String body = getBody(request);
        Map<String, String> headers = getHeaders(request);
        Map<String, String> params = new HashMap<>();
        if (method!= RequestModel.Method.GET && "application/x-www-form-urlencoded".equals(headers.get("Content-Type"))) {
            bodyToParam(body, params);
            body = null;
        }
        return new RequestModel(method, url, headers, params, body);
    }


    private static void bodyToParam(String body,Map<String, String> ps) {
        if (body == null) return;

        String[] kvs = body.trim().split("&");
        for (String kv : kvs) {
            String[] kvss = kv.split("=");
            if (kvss.length == 2) {
                String k = kvss[0].trim();
                String v = kvss[1].trim();
                ps.put(k, v);
            }
        }
    }

    private static Map<String, String> getHeaders(cn.vove7.plugin.rest.psi.RestRequest request) {
        Map<String, String> result = new HashMap<>();
        RestHeaders headers = request.getHeaders();
        if (headers != null) {
            for (RestEHeader header : headers.getEHeaderList()) {
                String text = header.getText();
                int i = text.indexOf("@");
                if (i >= 0) {
                    i = text.indexOf(":");
                    if (i >= 0) {
                        String name = text.substring(1, i).trim();
                        String value = text.substring(i + 1).trim();
                        result.put(name, value);
                    }
                }
            }
        }
        return result;
    }

    private static String getBody(cn.vove7.plugin.rest.psi.RestRequest request) {
        RestRequestBody body = request.getRequestBody();
        if (body != null && body.isValid()) {
            return body.getText();
        }
        return null;
    }

    @NotNull
    private static String getUrl(cn.vove7.plugin.rest.psi.RestRequest request) {
        RestEUrl url = request.getEUrl();
        if (url.isValid()) {
            String urlText = url.getText();
            boolean markExists = urlText.contains("?");
            StringBuilder sb = new StringBuilder(urlText);
            RestParams params = request.getParams();
            if (params != null && params.isValid()) {
                if (!markExists) {
                    sb.append('?');
                }
                for (RestEParam param : params.getEParamList()) {
                    sb.append(param.getText());
                }
            }
            return sb.toString();
        }
        throw new IllegalStateException("");
    }

}

File: src/main/kotlin/cn/vove7/plugin/rest/filetype/RequestParser.kt
package cn.vove7.plugin.rest.filetype

import cn.vove7.plugin.rest.model.RequestModel
import cn.vove7.plugin.rest.psi.RestRequest

/**
 * @author danblack
 */
object RequestParser {
    private fun getMethod(request: RestRequest): RequestModel.Method {
        val method = request.eMethod
        if (method != null && method.isValid) {
            when (method.text) {
                "GET" -> return RequestModel.Method.GET
                "PUT" -> return RequestModel.Method.PUT
                "POST" -> return RequestModel.Method.POST
                "DELETE" -> return RequestModel.Method.DELETE
                "PATCH" -> return RequestModel.Method.PATCH
            }
            throw IllegalStateException()
        }
        return RequestModel.Method.GET
    }

    fun parse(request: RestRequest): RequestModel {
        val method = getMethod(request)
        val url = getUrl(request)
        var body = getBody(request)
        val headers = getHeaders(request)
        val params: MutableMap<String, String> = HashMap()
        if (method !== RequestModel.Method.GET && "application/x-www-form-urlencoded" == headers["Content-Type"]) {
            bodyToParam(body, params)
            body = null
        }
        return RequestModel(method, url, headers, params, body)
    }

    private fun bodyToParam(body: String?, ps: MutableMap<String, String>) {
        if (body == null) return
        val kvs = body.trim { it <= ' ' }.split("&".toRegex()).toTypedArray()
        for (kv in kvs) {
            val kvss = kv.split("=".toRegex()).toTypedArray()
            if (kvss.size == 2) {
                val k = kvss[0].trim { it <= ' ' }
                val v = kvss[1].trim { it <= ' ' }
                ps[k] = v
            }
        }
    }

    private fun getHeaders(request: RestRequest): MutableMap<String, String> {
        val result: MutableMap<String, String> = HashMap()
        val headers = request.headers
        if (headers != null) {
            for (header in headers.eHeaderList) {
                val text = header.text
                var i = text.indexOf("@")
                if (i >= 0) {
                    i = text.indexOf(":")
                    if (i >= 0) {
                        val name = text.substring(1, i).trim { it <= ' ' }
                        val value = text.substring(i + 1).trim { it <= ' ' }
                        result[name] = value
                    }
                }
            }
        }
        return result
    }

    private fun getBody(request: RestRequest): String? {
        val body = request.requestBody
        return if (body != null && body.isValid) {
            body.text
        } else null
    }

    private fun getUrl(request: RestRequest): String {
        val url = request.eUrl
        if (url.isValid) {
            val urlText = url.text
            val markExists = urlText.contains("?")
            val sb = StringBuilder(urlText)
            val params = request.params
            if (params != null && params.isValid) {
                if (!markExists) {
                    sb.append('?')
                }
                for (param in params.eParamList) {
                    sb.append(param.text)
                }
            }
            return sb.toString()
        }
        throw IllegalStateException("")
    }
}